docs/pages/pmd/userdocs/cpd.md

```diff
@@ -30,6 +30,33 @@ See how to add it [here](/pmd_devdocs_adding_new_cpd_language.html).
 
 CPD is included with PMD, which you can download [here](http://sourceforge.net/projects/pmd/files/pmd/).
 
+### Why should you care?
+
+It's certainly important to know where to get CPD, and how to call it, but it's worth stepping back for a moment and asking yourself why you should care about this, being the occurrence of duplicate code blocks.
+
+#### If nothing ever changes, the impact is minimal
+
+If you're analyzing code that never changes, it's reasonable to assume that you don't need to care too much about this.  Risk comes in when you're working with complex code that can change over time.
+
+#### Refactoring known duplications can be messy
```

Comment:
Don't these sections basically say the same thing? I think your case may be summed up in a few bullet points:
* Assuming duplicated blocks of code are supposed to do the same thing, any refactoring, even simple, must be duplicated too -- which is unrewarding grunt work, and puts pressure on the developer to find every place in which to perform the refactoring.
  * Failure to keep the code in sync may mean automated tools will no longer recognise these blocks as duplicates. This means the task of finding duplicates to keep them in sync when doing subsequent refactorings can no longer be entrusted to an automated tool -- and humans do mistakes. Segments of code supposed to do the same thing may grow apart undetected upon further refactoring.
* If the code may never change in the future, then this is not a problem
* Otherwise, the most viable solution is to not duplicate. Possible ways to remove duplicates:
  * When the duplication is local to a method or single class: extract a local variable/ method
  * When the duplication occurs in siblings within a class hierarchy: extract a method and pull it up in the class hierarchy/ use the Template Method design pattern
   * When the duplication occurs in unrelated hierarchies: use the Strategy pattern  
   * ...?

Reply:
Which specific sections are you suggesting say the same thing?

I don't know that I would be willing to "sum up" these ideas.

Your mention of refactoring strategies is a good idea, but that calls for spelling out that information in text, not just summarizing in bullet points.

Reply:
These excerpts from the section about the first risk just states that the blocks need to be kept in sync without exactly explaining why it would be bad for them not to be dups anymore:

> you still have to make sure that you’ve found all the duplicated blocks and made the exact same change in each one, not just so they are logically correct, but so that they are still duplicate blocks.

> you have to ensure that they are still recognized as duplicate blocks,

At this point we could ask ourselves: why would we want them to stay duplicates? The next section clarifies that by describing consequences of syncing mistakes:

>  QUOTE A: *making it possible that the person doing the refactoring simply won’t know about that code that needs to change.* After that refactoring, you now have two blocks of code that were supposed to do the same thing, but now they are not. 

But it also kind of repeats what was stated in the first section, which adds noise IMHO:

> If you are not careful when refactoring one or more of the code blocks, you could accidentally leave them so they are no longer recognized as duplicates by the scanning technology, but they are still functionally identical.


In the third section:



> If someone makes logic changes to a block of code that has exact or logical duplicates somewhere else, while not considering those duplicates in any way, leads to the same situation at the end of the previous risk, being blocks that should have been doing the same thing, but now are not, and **it will be very difficult for anyone to know that**.


It seems to me that the bold part is the risk meant to be exposed by this sentence, correct? In that case, I think it was already stated in the previous section (italic part of quote A). For this reason, I don't really know how the third section explains a different scenario from the second. Would you care to explain about that?

---

IMHO the problem is that the fact they must be kept in sync is not a *risk* per se, it's a fact. The risk stems from syncing mistakes though, and is that *duplicates may not do the same thing after a while*. This is undesirable for many reasons, and aggravated by the fact that automated tools don't help anymore (which is not really a risk either).  

I don't intend the doc to be made in bullet points, I just used them to present another structure for this documentation, that avoids repetition. E.g. the first bullet point may be a section, and similarly for others.